File: ARCM/Python/odefun.py
import mindspore.numpy as np
from mindspore import Tensor
from mindspore.ops import operations as ops
import fka, fkb, Jacba, Jacbb, dJA, dJB, ddJB, fish_rd, get_T, get_dT, JEA, dJEA, get_M, get_B
from mindspore.ops import operations as Plus

def fun(state, t):
    state = Tensor(state, np.float32)
    theta = state[:7].reshape((7, 1))
    dtheta = state[7:14].reshape((7, 1))
    k = state[14]
    dk = state[15]
    d = state[16:20].reshape((4, 1))
    X = state[20:34].reshape((14, 1))
    ddtheta = X[:7]
    ddk = state[28]
    dcompensate = state[34:38].reshape((4, 1))
    ncompensate = state[38:52].reshape((14, 1))
    JeA = state[52:73].reshape(3, 7)
    P = state[73]
    
    alpha = 40
    beta = 10
    d0 = 50
    d1 = 10
    m1 = 0.1
    m2 = 0.1
    sigma = 100
    
    deta = np.array([50 + i * t for i in range(1, 5)], np.float32).reshape(4, 1)
    neta = np.array([50 + i * t for i in range(1, 15)], np.float32).reshape(14, 1)
    dgamma = 100
    dlambda = 100000
    ngamma = 100
    nlambda = 100000
    
    ra = fka.a(theta)
    rb = fkb.b(theta)
    rc = rb + k * (ra - rb)
    
    b = np.concatenate((ra, np.array([1], np.float32)), axis=0)
    
    Ja = Jacba.a(theta)
    Jb = Jacbb.b(theta)
    dJa = dJA.A(theta, dtheta)
    dJb = dJB.B(theta, dtheta)
    ddJb = ddJB.B(theta, dtheta, ddtheta)
    
    dddrd, ddrd, drd, rd = fish_rd.d(t)
    T = get_T.T(theta)
    dT = get_dT.T(theta, dtheta)

    dra = Ja @ Tensor(dtheta)
    drb = Jb @ dtheta
    ddra = dJa @ dtheta + Ja @ ddtheta
    db = np.concatenate((dra, np.array([0], np.float32)), axis=0)
    
    # nNSND
    dJeA = 10000 * (JEA.A(theta, d) - JeA)
    ddJeA = 100 * (dJEA.A(theta, dtheta, d) - dJeA)
    
    M, dM = get_M.M(JeA, dJeA, Jb, dJb, k, dk, ra, dra, rb, drb)
    B, dB = get_B.B(theta, dtheta, ddtheta, k, dk, ddk, JeA, dJeA, ddJeA, Jb, dJb, ddJb, rd, drd, ddrd, dddrd, ra, dra, ddra, rc)
    
    # dNSND
    dX = ops.linalg.pinv(M) * (dB - ops.matmul(dM, X) - nlambda * (ops.matmul(M, X) - B) - ngamma * ((ops.matmul(M, X) - B) + ngamma * nlambda * ncompensate)) + 100 * neta
    dncompensate = ops.matmul(M, X) - B
    
    dd = ops.linalg.pinv(T)*(db - ops.matmul(dT, d) - dlambda*(ops.matmul(T, d) - b) - dgamma*((T*d - b) + dgamma*dlambda*dcompensate)) + 100*deta
    ddcompensate = ops.matmul(T, d) - b
    
    # Kalman
    F = 1
    G = 1
    H = 1
    Q = 0.01
    R = 0.25
    W = Tensor(np.sqrt(Q) * np.random.randn(14, 1), np.float32)
    V = Tensor(np.sqrt(R) * np.random.randn(14, 1), np.float32)

    # 计算 K
    K = ops.matmul(ops.matmul(P, H), ops.linalg.pinv(R))

    # 计算 dX
    dX = F * dX + G * W + ops.matmul(K, V)

    # 计算 dP
    dP = F * P + ops.matmul(P, ops.transpose(F, (1, 0))) + Q - ops.matmul(ops.matmul(K, R), ops.transpose(K, (1, 0)))
    
    dstate = np.concatenate((dtheta.flatten(), ddtheta.flatten(), np.array(dk).flatten(), np.array(ddk).flatten(), dd.flatten(), dX.flatten(), ddcompensate.flatten(), dncompensate.flatten(), dJeA.reshape(21, 1).flatten(), np.array(dP).flatten()))
    
    print(t)
    
    return dstate

File: ARCM/Matlab/odefun.m
function dstate = odefun(t, state)
 
% global init_theta init_rc %#ok<GVMIS> 
global alpha beta m1 m2 d0 d1 sigma %#ok<GVMIS> 

theta = state(1:7);   dtheta = state(8:14);
k = state(15);   dk = state(16);
d = state(17:20);
X = state(21:34);
ddtheta = X(1:7);   ddk = X(8);   % lambda1 = X(9:11);   lambda2 = X(12:14);
dcompensate = state(35:38);
ncompensate = state(39:52);
JeA = reshape(state(53:73), 3, 7);
P = state(74);

ddtheta = qLimit(ddtheta);

% zeta = 1e5;
alpha = 40;   beta = 10;
d0 = 50;   d1 = 10;   m1 = 0.1;   m2 = 0.1;
sigma = 100;
deta = [50+1*t;50+2*t;50+3*t;50+4*t];
neta = [50+1*t;50+2*t;50+3*t;50+4*t;50+5*t;50+6*t;50+7*t;50+8*t;50+9*t;50+10*t;50+11*t;50+12*t;50+13*t;50+14*t];
dgamma = 1e2;   dlambda = 1e5;
ngamma = 1e2;   nlambda = 1e5;

ra = fka(theta);   b = [ra; 1];
rb = fkb(theta);
rc = rb + k*(ra - rb);

Ja = Jacba(theta);   Jb = Jacbb(theta);
dJa = dJA(theta, dtheta);   dJb = dJB(theta, dtheta);
ddJb = ddJB(theta, dtheta, ddtheta);
[dddrd, ddrd, drd, rd]=fish_rd(t);
T = get_T(theta);   dT = get_dT(theta, dtheta);

dra = Ja*dtheta;   db = [dra; 0];
drb = Jb*dtheta;
ddra = dJa*dtheta + Ja*ddtheta;

% 原程序
% W = k*JeA + (1 - k)*Jb;
% dJeA = 10000*(JEA(theta, d) - JeA);
% dddtheta = -zeta*(ddtheta + (alpha*dtheta + beta*(theta - init_theta)) + JeA'*lambda1 + W'*lambda2);
% dddk = -zeta*(m1*dk + m2*ddk + lambda2'*(ra - rb));
% dlambda1 = zeta*(JeA*ddtheta - ddrd + dJeA*dtheta - d0*(rd - ra) - d1*(drd - dra));
% dlambda2 = zeta*(W*ddtheta + (2*dk*(JeA - Jb) + k*dJeA + (1 - k)*dJb)*dtheta + ddk*(ra - rb) + sigma*(rc - init_rc));

% nNSND
dJeA = 10000*(JEA(theta, d) - JeA);
ddJeA = 100*(dJEA(theta, dtheta, d) - dJeA);
[M, dM] = get_M(JeA, dJeA, Jb, dJb, k, dk, ra, dra, rb, drb);
[B, dB] = get_B(theta, dtheta, ddtheta, k, dk, ddk, JeA, dJeA, ddJeA, Jb, dJb, ddJb, rd, drd, ddrd, dddrd, ra, dra, ddra, rc);
% parameter of end-effetor
dX = pinv(M)*(dB - dM*X - nlambda*linear(M*X - B) - ngamma*(linear(M*X - B) + ngamma*nlambda*ncompensate)) + 100*neta;
dncompensate = linear(M*X - B);

% dNSND
% noise-supperising during working
dd = pinv(T)*(db - dT*d - dlambda*linear(T*d - b) - dgamma*(linear(T*d - b) + dgamma*dlambda*dcompensate)) + 100*deta;
ddcompensate = linear(T*d - b);

% Kalman filter
F = 1;
G = 1;
H = 1;
Q = 0.01;
R = 0.25;
W = sqrt(Q)*randn(14, 1);
V = sqrt(R)*randn(14, 1);

K = P*H*pinv(R);
dX = F*dX + G*W + K*V;
dP = F*P + P*F' + Q - K*R*K';


dstate = [dtheta; ddtheta; dk; ddk; dd; dX; ddcompensate; dncompensate; reshape(dJeA, 21, 1); dP];

disp(t)
end
